### PR TITLE
[TT-5117] Implemented duplicate Input/Object/Interface removal

### PR DESCRIPTION
**Pull request by Aenimus**
[TT-5117](https://tyktech.atlassian.net/browse/TT-5117)

### CODE REVIEW

File: pkg/federation/sdlmerge/sdlmerge.go [hunk 1]
```diff
@@ -99,10 +99,11 @@ func (m *normalizer) setupWalkers() {
 		// visitors for cleaning up federated duplicated fields and directives
 		{
 			newRemoveFieldDefinitions("external"),
+			newRemoveDuplicateFieldedSharedTypesVisitor(),
+			newRemoveDuplicateFieldlessSharedTypesVisitor(),
 			newRemoveInterfaceDefinitionDirective("key"),
 			newRemoveObjectTypeDefinitionDirective("key"),
```
Comment (Aenimus) [inline pkg/federation/sdlmerge/sdlmerge.go:105]: Looks like the handler for removing Input directives is missing(?)

File: pkg/federation/sdlmerge/remove_duplicate_fielded_shared_types.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	r.lastObjectRef = ref
+}
+
+func (r *removeDuplicateFieldedSharedTypesVisitor) LeaveDocument(_, _ *ast.Document) {
+	if r.rootNodesToRemove != nil {
+		r.document.DeleteRootNodes(r.rootNodesToRemove)
+	}
+}
+
+func (r *removeDuplicateFieldedSharedTypesVisitor) checkForDuplicateEntity(hasDirectives bool, directiveRefs []int, name string) {
```
Comment (Aenimus) [inline pkg/federation/sdlmerge/remove_duplicate_fielded_shared_types.go:115]: This is the logic that handles entities.

File: pkg/federation/sdlmerge/remove_duplicate_fielded_shared_types.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+func (r *removeDuplicateFieldedSharedTypesVisitor) EnterInputObjectTypeDefinition(ref int) {
+	if ref <= r.lastInputRef {
+		return
+	}
+	document := r.document
+	name := document.InputObjectTypeDefinitionNameString(ref)
+	inputType := document.InputObjectTypeDefinitions[ref]
+	r.checkForDuplicateEntity(inputType.HasDirectives, inputType.Directives.Refs, name)
```
Comment (Aenimus) [inline pkg/federation/sdlmerge/remove_duplicate_fielded_shared_types.go:53]: Each entry into a fielded type now calls this method (defined below). The method checks whether the fielded type is an entity, and if it is, adds it to the entity set. If the entity name already exists in the set, we know it's duplicated, and the federation is immediately invalid.

File: pkg/federation/sdlmerge/remove_duplicate_fielded_shared_types.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+func (r *removeDuplicateFieldedSharedTypesVisitor) checkForDuplicateEntity(hasDirectives bool, directiveRefs []int, name string) {
+	if r.entitySet[name] {
+		r.StopWithExternalErr(operationreport.ErrEntitiesMustNotBeSharedTypes(name))
+	}
+	if !hasDirectives {
+		return
+	}
+	for _, directiveRef := range directiveRefs {
+		if r.document.DirectiveNameString(directiveRef) == "key" {
+			if _, exists := r.sharedTypeSet[name]; exists {
```
Comment (Aenimus) [inline pkg/federation/sdlmerge/remove_duplicate_fielded_shared_types.go:124]: This logic was necessary if the entity occurs at the end.

File: pkg/federation/sdlmerge/remove_duplicate_fielded_shared_types.go (hunk 3 above)
Comment (pvormste) [reply pkg/federation/sdlmerge/remove_duplicate_fielded_shared_types.go:53]: only object types and interfaces can be entitites (a.k.a. have a `@key` directive)

See: https://www.apollographql.com/docs/federation/v1/federation-spec
```graphql
directive @key(fields: _FieldSet!) repeatable on OBJECT | INTERFACE
```

**Review by pvormste — CHANGES_REQUESTED** [review]

File: pkg/federation/sdlmerge/remove_duplicate_fielded_shared_types.go (hunk 3 above)
Comment (Aenimus) [reply pkg/federation/sdlmerge/remove_duplicate_fielded_shared_types.go:53]: Of course. And that's why there is not visitor. Sorry.

If I remove this, are you happy with the implementation?

**Review by buraksezer — APPROVED** [review]

**Pull request merged**